app/models/workers_scaling/algorithm_factory.rb

```diff
@@ -63,69 +63,41 @@ def self.create_algorithm(attributes)
     end
 
     ##
-    # Arguments:
-    #  * experiment_id - id of Experiment to be subjected to Algorithm
-    #  * user_id - id of User starting Algorithm
-    #  * allowed_infrastructures - list of hashes with infrastructure and maximal Workers amount
-    #      (Detailed description at ExperimentResourcesInterface#initialize)
-    #  * planned_finish_time - planned time of Experiment end
-    #  * params - workers scaling params passed from ExperimentsController
-    def self.initial_deployment(experiment_id, user_id, allowed_infrastructures, planned_finish_time, params)
-      algorithm = create_algorithm(
-          class_name: params[:name].to_sym,
-          experiment_id: experiment_id,
-          user_id: user_id,
-          allowed_infrastructures: allowed_infrastructures,
-          planned_finish_time: planned_finish_time,
-          last_update_time: Time.now
-      )
-      algorithm.save
-      Thread.new do
-        begin
-          algorithm.initial_deployment
-          algorithm.notify_execution
-          LOGGER.debug 'Initial deployment finished'
-        rescue => e
-          LOGGER.error "Exception occurred during initial deployment: #{e.to_s}\n#{e.backtrace.join("\n")}"
-          raise
-        end
-      end
-    end
-
-
-    ##
-    # Returns instance of algorithm from cache
+    # Returns instance of algorithm from cache, nil if algorithm does not exist
     # If Algorithm is not stored in cache, new one will be created
     # If stored algorithm has last_update_time older than the one in database,
     # it will be deleted from cache
+    # @param experiment_id [BSON::ObjectId, String]
+    # @return [class < Algorithm, nil]
     def self.get_algorithm(experiment_id)
       cache_key = "workers_scaling_algorithm_#{experiment_id}"
       if @cache.exist?(cache_key)
-        last_update_time = Algorithm.where({experiment_id: experiment_id},
-                                           fields: [:last_update_time])
-                               .first.last_update_time
-        if @cache.read(cache_key).last_update_time < last_update_time
+        algorithm = Algorithm.where({experiment_id: experiment_id}, fields: [:last_update_time]).first
+        return nil if algorithm.blank?
+        if @cache.read(cache_key).last_update_time < algorithm.last_update_time
```

Comment:
Jesli nie bylo robionych testow to nie wiemy na ile o ile szkodliwa byla implementacja bez cache'a. Nie wiemy tez na ile i o ile wprowadzenie cache'a poprawilo ten stan. Jest to w.g. mnie przedwczesna optymalizacja ktora jak wiadomo jest szkodliwa.